### DISCUSSION TITLE
How do I disable ruff for a specific line?

### DISCUSSION THREAD

**Discussion by atherfawaz** [astral-sh/ruff#3442] [Q&A] [answered]
I was wondering if there is a way to disable lint check/fix for one line of code. We have [similar options](https://stackoverflow.com/questions/28829236/is-it-possible-to-ignore-one-single-specific-line-with-pylint) in pylint. 

For example, I would like to do something like:
```python3
# ruff: disable=F401
from abc.xyz import function_name
```

Apologies if this has been answered previously; I couldn't find it in the Discussions.

**Comment by charliermarsh** [comment, marked as answer]
The most common would be to add a `noqa` inline, like so:

```py
from abc.xyz import function_name  # noqa: F401
```

If you want to ignore _all_ errors on a line:

```py
from abc.xyz import function_name  # noqa
```

You can also disable all errors of a given kind across an entire file like so:

```py
# ruff: noqa: F401

from abc.xyz import function_name
```

(More details in the [docs](https://beta.ruff.rs/docs/configuration/#error-suppression).)

**Reply by atherfawaz** [reply]
Thank you so much, this worked for me! 😄

**Reply by matu3ba** [reply]
> `# ruff: noqa: E501`

regressed or is not working for file-global specific errors. [as of ruff 0.0.243]

I do use `ruff --quiet --ignore E501 -` as workaround.

**Reply by charliermarsh** [reply]
I just checked and this is working correctly for me on main via `ruff check foo.py --isolated -n`:

```python
# ruff: noqa: E501

x = "ajksdasjkld;hasdoiasd ajksdasjkld;hasdoiasd ajksdasjkld;hasdoiasd ajksdasjkld;hasdoiasd ajksdasjkld;hasdoiasd ajksdasjkld;hasdoiasd"
```

Can you share a reproducible test case, and/or try again on main?

**Reply by matu3ba** [reply]
Oh, I see whats wrong now. My bad. My ruff version is way too old on my work pc. On my home pc it shows:
```
ruff --version
ruff 0.0.269
```

**Reply by LumenYoung** [reply]
> Oh, I see whats wrong now. My bad. My ruff version is way too old on my work pc. On my home pc it shows:
> 
> ```
> ruff --version
> ruff 0.0.269
> ```

Hi, I'm also experiencing this. I'm installing the ruff from mason. but I don't think this version is too old right? Because it is the latest release on the ruff github page. How did you get this resolved?

**Reply by fraunhofersociety** [reply]
I suggested to support rule-names instead of rule-codes, too:
https://github.com/astral-sh/ruff/issues/15315

**Comment by bersbersbers** [comment]
The difference between `# ruff: noqa: XXX` and `noqa: XXX` is really subtle. While you would think that the difference is that the former applies to only one tool while the latter applies to more, the difference actually is the scope (file vs line). That is not extremely intuitive ;)

**Reply by charliermarsh** [reply]
I agree. This is something we borrowed from Flake8, which has the same system. We'll likely redo suppression comments entirely at some point.

**Comment by rkshthrmsh** [comment]
Hi @charliermarsh, is there a way to disable ruff for a block of lines? `# noqa` and `# ruff: noqa` work for a single line and an entire file, respectively. Would this require commenting `# noqa` on all lines in the block?

**Reply by charliermarsh** [reply]
Ah yeah, there isn’t right now. We’ll likely support it in the future, but the syntax is TBD. So you’re right: you either need to disable for the file, or comment on every line. (I’d suggest the former.)

**Reply by k4lizen** [reply]
@charliermarsh just checking in, still no way to do this?

**Reply by MichaReiser** [reply]
See https://github.com/astral-sh/ruff/issues/3711

**Comment by bwagner** [comment]
I only want to import `typer` if I'm running a module as a script:
```
if __name__ == "__main__":
    import typer  # noqa: E402
```
However, running this through `ruff` still outputs the error:
`script.py:26:1: E402 Module level import not at top of file`

python v3.11.4
macOS 13.4.1
ruff 0.0.282

**Reply by charliermarsh** [reply]
Are you able to post a complete file for reproduction? (The error says line 26.)

**Reply by zanieb** [reply]
(We should probably not apply that rule in `if __name__ == "__main__":` blocks)

**Comment by bwagner** [comment]
Here is a reduced script that triggers the ruff E402 error. Note the
comments "delete this". Delete those lines individually and adapt main to
circumvent syntax errors: Each deletion on its own weirdly make the E402
error disappear.

import sys  # delete this and adapt main -> ruff error E402 disappears

"Delete this comment and the ruff error E402 disappears."

from pathlib import Path  # delete this and adapt main -> ruff error E402
disappears

if __name__ == "__main__":
    import os  # noqa: E402

    def main():
        print(Path(sys.argv[0]))
        print(os.name)

    main()

**Comment by T4mmi** [comment]
Hi, sorry to dig it up, but is there any way to properly skip a single line for ANY syntax error ?  my problem is that I use ruff for a cookiecutter template wich comes with a pytest case importing the package :
```py
import {{cookiecutter.project_slug}} as pkg # noqa ???
```
this will trigger a
```sh
invalid-syntax: Expected a statement
 --> {{ cookiecutter.project_name }}\tests\test_package.py:1:38
```
that I cannot remove ...  
I tried to add this to the `ruff.toml` but had no luck with it
```toml
exclude = ["** cookiecutter.project_name **/tests/test_package.py"]
```

**Reply by MichaReiser** [reply]
Syntax errors can't be suppressed by noqa comments. The recommended way is to exclude the cookie cutter files as shown here https://github.com/astral-sh/ruff/issues/7959#issuecomment-1764751734

**Reply by T4mmi** [reply]
thanks for the reply ... but this approach isn't working either :(

**Reply by MichaReiser** [reply]
Can you share your exclude pattern and the exact paths you want to exclude. You may also want to run ruff with `ruff check -v` to see why it excludes (or doesn't) certain files

**Reply by T4mmi** [reply]
i have something like that:
```
|- ruff.toml (1)
|- {{ cookiecutter.project_name }}
.  |- ruff.toml (2)
.  |- tests
.  .  |- test_package.py
.  |- src
.  .  |- {{ cookiecutter.project_name }}
.  .  .  |- __init__.py
```
with `ruff.toml (1)`:
```toml
[lint]
exclude = [
    "** cookiecutter.project_name **/tests/test_package.py",
    "** cookiecutter.project_name **/src/** cookiecutter.project_slug **/__init__.py",
]
```

**Reply by T4mmi** [reply]
ok the pb is that the second `ruff.toml (2)` will take precedence over the (1) ! running `ruff --config ruff.toml` fixes the issue (see https://github.com/astral-sh/ruff/issues/9585#issuecomment-3431607789)

**Reply by MichaReiser** [reply]
Yeah, you want to either exclude the entire directory (in which case ruff doesn't traverse into it) or explicitly set the configuration
